RFR - Drf Implement Index

davidrf opened the pull request.

davidrf committed:
Add migrations, models, controllers

davidrf committed:
Added models, schema

davidrf committed:
Update master

davidrf committed:
Got index page working

davidrf committed:
Index page with changes

davidrf committed:
Index page with final changes

vikram7 commented on app/controllers/beers_controller.rb:
```diff
@@ -0,0 +1,51 @@
+      flash[:notice] = "Beer was saved"
+      redirect_to 'beer#show'
+    else
+      render 'new'
```
add a notice for a failure

vikram7 commented on app/controllers/beers_controller.rb:
```diff
@@ -0,0 +1,51 @@
+
+  def destroy
+    Beer.delete(params[:id])
+    flash[:notice] = "Beer deleted."
```
What if there is a failure?

vikram7 commented on app/controllers/reviews_controller.rb:
```diff
@@ -0,0 +1,46 @@
+      flash[:notice] = "Review Saved"
+      redirect_to beer_show_path(@beer)
+    else
+      render :new
```
add notice for failure

vikram7 commented on app/controllers/reviews_controller.rb:
```diff
@@ -0,0 +1,46 @@
+  def update
+    @review = Review.find(params[:id])
+    @beer = Beer.find(params[:id])
+    if @Review.update(review_params)
```
lowercase `@review`

vikram7 replied:
do you mean lowercase `@review`?

vikram7 commented on config/routes.rb:
```diff
@@ -1,5 +1,6 @@
+  root 'beers#index'
   devise_for :users
 
+  resources :beers, only: [:index]
```
if you have methods in your reviews controller, you need to update your routes accordingly. Would you need nested beer/review routes?

davidrf committed:
Updated destroy routes

vikram7 commented on app/controllers/beers_controller.rb:
```diff
@@ -0,0 +1,55 @@
+    if @beer.destroy
+      flash[:notice] = "Beer deleted."
+    else
+      flash[:notice] = "Beer was not deleted"
```
Minor point. Line 45 does not end in a period and line 43 does. Make sure they are all consistent with other controllers.

davidrf committed:
Revise according to VR commets

davidrf committed:
Revise according to VR comments 2

davidrf committed:
Downcase database name

Testing to see if this works with Travis CI. Previously it could
not find the database even if it was configured in the .travis.yml
file.

davidrf committed:
Comment tests out

atsheehan commented on app/controllers/beers_controller.rb:
```diff
@@ -0,0 +1,55 @@
+  end
+
+  def create
+    @beer = Beer.new(beer_params)
```
Where is the user authentication? How do you prevent unauthorized users from accessing this action?

``` ruby
class BeersController < ApplicationController
  before_action :authenticate_user!, except: [:index, :show]

  def new
    @beer = Beer.new
  end

  def create
    @beer = Beer.new(beer_params)
    @beer.user = current_user

    if @beer.save
    # ...
```

atsheehan commented on app/controllers/beers_controller.rb:
```diff
@@ -0,0 +1,55 @@
+    @beer = Beer.new(beer_params)
+    if @beer.save
+      flash[:notice] = "Beer was saved"
+      redirect_to 'beer#show'
```
This might need a path instead:

``` ruby
redirect_to beer_path(@beer)
redirect_to @beer
```

atsheehan commented on app/controllers/beers_controller.rb:
```diff
@@ -0,0 +1,55 @@
+  end
+
+  def edit
+    @beer = Beer.find(params[:id])
```
This will allow any user to edit any beer. To scope to beers that I own:

``` ruby
def edit
  @beer = Beer.find_by!(user: current_user, id: params[:id])
  # alternatively
  @beer = current_user.beers.find(params[:id])
end
```

atsheehan commented on app/controllers/beers_controller.rb:
```diff
@@ -0,0 +1,55 @@
+  end
+
+  def update
+    @beer = Beer.find(params[:id])
```
Same as above, limit this to the current user's beers.

atsheehan commented on app/controllers/beers_controller.rb:
```diff
@@ -0,0 +1,55 @@
+  end
+
+  def destroy
+    @beer = Beer.find(params[:id])
```
Same as above, only allow user to destroy their own beers.

atsheehan commented on app/controllers/reviews_controller.rb:
```diff
@@ -0,0 +1,50 @@
+
+  def create
+    @review = Review.new(review_params)
+    @beer = Beer.find(params[:beer_id])
```
How are you associating this review with the beer?

``` ruby
def create
  @beer = Beer.find(params[:beer_id])
  @review = Review.new(review_params)
  @review.beer = @beer
  @review.user = current_user

  if @review.save
    # ...
```

atsheehan commented on app/controllers/reviews_controller.rb:
```diff
@@ -0,0 +1,50 @@
+  end
+
+  def update
+    @review = Review.find(params[:id])
```
Edit and update should scope to the user's own reviews.

atsheehan commented on app/controllers/reviews_controller.rb:
```diff
@@ -0,0 +1,50 @@
+
+  def update
+    @review = Review.find(params[:id])
+    @beer = Beer.find(params[:beer_id])
```
You don't need to find the beer explicitly. You can find the beer from the review object:

``` ruby
@beer = @review.beer
```

atsheehan commented on app/controllers/reviews_controller.rb:
```diff
@@ -0,0 +1,50 @@
+  protected
+
+  def review_params
+    params.require(:review).permit(:rating, :description, :beer_id, :user_id)
```
Don't accept `:user_id` through the params. This would allow a user to impersonate someone else just by changing their ID to some arbitrary value.

The `beer_id` is passed via the URL (`/beers/10/reviews`) so you can extract the ID there rather than through the form.

atsheehan commented on app/models/beer.rb:
```diff
@@ -0,0 +1,8 @@
+  belongs_to :user
+  validates :name, presence: true
+  validates :style_id, presence: true
+  validates :user_id, presence: true
```
Validate the association rather than the ID:

``` ruby
validates :style, presence: true
```

atsheehan commented on app/models/review.rb:
```diff
@@ -0,0 +1,8 @@
+  validates :rating, presence: true
+  validates :description, presence: true
+  validates :user_id, presence: true
+  validates :beer_id, presence: true
```
Validate on the association rather than ID.

atsheehan commented on app/models/review.rb:
```diff
@@ -0,0 +1,8 @@
+  belongs_to :beer
+  belongs_to :user
+  validates :rating, presence: true
+  validates :description, presence: true
```
Does description have a max length?

atsheehan commented on app/models/review.rb:
```diff
@@ -0,0 +1,8 @@
+class Review < ActiveRecord::Base
+  belongs_to :beer
+  belongs_to :user
+  validates :rating, presence: true
```
Is rating a numeric value? What is the acceptable range?

atsheehan commented on app/models/style.rb:
```diff
@@ -0,0 +1,4 @@
+class Style < ActiveRecord::Base
+  has_many :beers
+  validates :name, presence: true
```
Should this name be unique?

atsheehan commented on config/routes.rb:
```diff
@@ -1,5 +1,8 @@
   devise_for :users
 
+  resources :beers do
+    resources :reviews, except: [:index, :show]
```
The edit, update, and destroy actions don't need to be nested.

``` ruby
resources :beers do
  resources :reviews, only: [:new, :create] # maybe index
end

resources :reviews, only: [:edit, :update, :destroy, :show] # maybe index
```

atsheehan commented on db/migrate/20150331170121_create_beers.rb:
```diff
@@ -0,0 +1,12 @@
+      t.string :name, null: false
+      t.integer :abv
+      t.integer :ibu
+      t.text :description
```
Add a length limit to the description:

``` ruby
t.text :description, length: 5000
```

atsheehan commented on db/migrate/20150331170121_create_beers.rb:
```diff
@@ -0,0 +1,12 @@
+    create_table :beers do |t|
+      t.string :name, null: false
+      t.integer :abv
+      t.integer :ibu
```
Maybe convert these to floats

atsheehan commented on db/migrate/20150331170609_create_styles.rb:
```diff
@@ -0,0 +1,7 @@
+class CreateStyles < ActiveRecord::Migration
+  def change
+    create_table :styles do |t|
+      t.string :name, null: false
```
Maybe add a unique index on this.

atsheehan commented on db/migrate/20150331170121_create_beers.rb:
```diff
@@ -0,0 +1,12 @@
+      t.integer :ibu
+      t.text :description
+      t.integer :style_id, null: false
+      t.integer :user_id, null: false
```
If you're going to be searching for beers by user or beers by style, add an index on each foreign key to speed up the queries.

atsheehan commented on db/migrate/20150331170704_create_reviews.rb:
```diff
@@ -0,0 +1,11 @@
+      t.integer :rating, null: false
+      t.text :description, null: false
+      t.integer :user_id, null: false
+      t.integer :beer_id, null: false
```
Add indexes on these foreign keys to speed up queries for reviews by user or reviews by beer.

atsheehan commented on spec/features/admin_deletes_a_beer_spec.rb:
```diff
@@ -1,32 +1,32 @@
+#     click_link beer.name
+#     expect(page).to_not have_content("Delete Beer")
+#   end
+# end
```
Delete commented out code (or uncomment).

atsheehan commented on spec/support/factory_girl.rb:
```diff
@@ -7,4 +7,23 @@
+  factory :beer do
+    name "Magic Hat"
+    abv "5"
+    ibu "60"
```
These are integers.

atsheehan commented on spec/support/factory_girl.rb:
```diff
@@ -7,4 +7,23 @@
+  end
+
+  factory :style do
+    name "IPA"
```
Maybe use sequences for unique factory records:

``` ruby
factory :style do
  sequence(:name) { |n| "IPA #{n}" }
end
```
